spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.utils
+
+import java.util.concurrent.{ConcurrentHashMap, CyclicBarrier}
+
+import com.intel.analytics.bigdl.tensor.Tensor
+import org.apache.spark.TaskContext
+
+import scala.reflect._
+
+/**
+ * A parameter synchronizer among threads per task
+ */
+object ParameterSynchronizer {
+
+  val fEvents = new java.util.concurrent.ConcurrentHashMap[String, Event[Float]]
+
+  val dEvents = new java.util.concurrent.ConcurrentHashMap[String, Event[Double]]
+
+  /**
+   * Register event with key and total thread number
+   * @param eventKey key to specify an event
+   * @param threadNum total thread number synchronizing on this key
+   * @tparam T
+   */
+  def register[T: ClassTag](eventKey: String, threadNum: Int): Unit = {
+    if (classTag[T] ==  classTag[Float]) {
+      var event = fEvents.get(eventKey)
+      if (event == null) {
+        event = new Event[Float](threadNum)
+        fEvents.putIfAbsent(eventKey, event)
+      }
+    } else if (classTag[T] ==  classTag[Double]) {
+      var event = dEvents.get(eventKey)
+      if (event == null) {
+        event = new Event[Double](threadNum)
+        dEvents.putIfAbsent(eventKey, event)
+      }
```

Comment:
else throw exception?

Reply:
we cannot throw exception here ,  it's for race condition and make sure the thread safety